src/sundials/sundials_adaptcontroller.c

```diff
@@ -0,0 +1,173 @@
+/* -----------------------------------------------------------------
+ * Programmer(s): Daniel R. Reynolds @ SMU
+ * -----------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2023, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------
+ * This is the implementation file for a generic SUNAdaptController
+ * package. It contains the implementation of the SUNAdaptController
+ * operations listed in sundials_adaptcontroller.h
+ * -----------------------------------------------------------------*/
+
+#include <sundials/sundials_adaptcontroller.h>
+
+/* -----------------------------------------------------------------
+ * Create a new empty SUNAdaptController object
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController SUNAdaptController_NewEmpty(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptController_Ops ops;
+
+  /* a context is required */
+  if (sunctx == NULL) return(NULL);
+
+  /* create controller object */
+  C = NULL;
+  C = (SUNAdaptController) malloc(sizeof *C);
+  if (C == NULL) return(NULL);
+
+  /* create matrix ops structure */
+  ops = NULL;
+  ops = (SUNAdaptController_Ops) malloc(sizeof *ops);
+  if (ops == NULL) { free(C); return(NULL); }
+
+  /* initialize operations to NULL */
+  ops->gettype              = NULL;
+  ops->destroy              = NULL;
+  ops->reset                = NULL;
+  ops->estimatestep         = NULL;
+  ops->setdefaults          = NULL;
+  ops->write                = NULL;
+  ops->seterrorbias         = NULL;
+  ops->updateh              = NULL;
+  ops->space                = NULL;
+
+  /* attach ops and initialize content to NULL */
+  C->ops     = ops;
+  C->content = NULL;
+  C->sunctx  = sunctx;
+
+  return(C);
+}
+
+
+/* -----------------------------------------------------------------
+ * Free a generic SUNAdaptController (assumes content is already empty)
+ * ----------------------------------------------------------------- */
+
+void SUNAdaptController_FreeEmpty(SUNAdaptController C)
```

Comment:
This function is missing from `sundials_adaptcontroller.h`
```suggestion
void SUNAdaptController_DestroyEmpty(SUNAdaptController C)
```